sup3r/bias/qdm.py

```diff
@@ -495,3 +501,341 @@ def run(self,
+            "sum" (daily sum/total)
+
+        Returns
+        -------
+        out : dict
+            Dictionary with parameters defining the statistical distributions
+            for each of the three given datasets. Each value has dimensions
+            (lat, lon, n-parameters).
+        """
+        logger.debug('Calculate CDF parameters for QDM')
+
+        logger.info(
+            'Initialized params with shape: {}'.format(
+                self.bias_gid_raster.shape
+            )
+        )
+        self.bad_bias_gids = []
+
+        # sup3r DataHandler opening base files will load all data in parallel
+        # during the init and should not be passed in parallel to workers
+        if isinstance(self.base_dh, DataHandler):
+            max_workers = 1
+
+        if max_workers == 1:
+            logger.debug('Running serial calculation.')
+            for i, bias_gid in enumerate(self.bias_meta.index):
+                raster_loc = np.where(self.bias_gid_raster == bias_gid)
+                _, base_gid = self.get_base_gid(bias_gid)
+
+                if not base_gid.any():
+                    self.bad_bias_gids.append(bias_gid)
+                    logger.debug(
+                        f'No base data for bias_gid: {bias_gid}. '
+                        'Adding it to bad_bias_gids'
+                    )
+                else:
+                    bias_data = self.get_bias_data(bias_gid)
+                    bias_fut_data = self.get_bias_data(
+                        bias_gid, self.bias_fut_dh
+                    )
+                    single_out = self._run_single(
+                        bias_data,
+                        bias_fut_data,
+                        self.base_fps,
+                        self.bias_feature,
+                        self.base_dset,
+                        base_gid,
+                        self.base_handler,
+                        daily_reduction,
+                        self.decimals,
+                        sampling=self.sampling,
+                        n_samples=self.n_quantiles,
+                        log_base=self.log_base,
+                        base_dh_inst=self.base_dh,
+                        zero_rate_threshold=zero_rate_threshold,
+                        bias_ti = self.bias_fut_dh.time_index,
+                        bias_fut_ti = self.bias_fut_dh.time_index,
+                    )
+                    for key, arr in single_out.items():
+                        self.out[key][raster_loc] = arr
+
+                logger.info(
+                    'Completed bias calculations for {} out of {} '
+                    'sites'.format(i + 1, len(self.bias_meta))
+                )
+
+        else:
+            logger.debug(
+                'Running parallel calculation with {} workers.'.format(
+                    max_workers
+                )
+            )
+            with ProcessPoolExecutor(max_workers=max_workers) as exe:
```

Comment:
I suppose this is an example of the duplicate code you mentioned. We have a lot of blocks with this kind of pattern, in completely independent classes / modules, and I've often wondered if it could be replaced with a custom decorator or something. Not worth doing right now though.

Reply:
Yes, fully agree. It would be nice to combine in one generic procedure. So far, I've been trying to follow the same patterns in the rest of the library.

I'm glad to work on that here or in a later PR. Your call.

Reply:
Missed the response here. Well put this on the todo list after the refactor integration.

Reply:
Issue #227 so we don't forget about it.